packages/@lwc/engine-core/src/framework/base-lightning-element.ts

```diff
@@ -554,9 +555,26 @@ LightningElement.prototype = {
         return vm.cmpRoot;
     },
 
-    get shadowRoot(): null {
-        // From within the component instance, the shadowRoot is always reported as "closed".
-        // Authors should rely on this.template instead.
+    get shadowRoot(): ShadowRoot | null {
+        const vm = getAssociatedVM(this);
+        const {
+            def: { ctor },
+        } = vm;
+        const { preferNativeShadow } = ctor;
+
+        if (process.env.NODE_ENV !== 'production') {
+            if (vm.renderMode === RenderMode.Light) {
+                logError(
+                    '`this.shadowRoot` returns null for light DOM components. Since there is no shadow, the rendered content can be accessed via `this` itself. e.g. instead of `this.shadowRoot.querySelector`, use `this.querySelector`.'
+                );
+            }
+        }
+
+        if (vm.renderMode !== RenderMode.Light && preferNativeShadow) {
```

Comment:
I suppose I technically don't need the `vm.renderMode !== RenderMode.Light` check here, since `vm.cmpRoot` is null for light DOM components, but I kind of like being explicit about it?